File: 47. Permutations II.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        Arrays.sort(nums);

        solve(res, nums, new ArrayList<>(), new boolean[nums.length]);

        return res;
    }

    public void solve(List<List<Integer>> res, int[] nums, List<Integer> path, boolean used[]){
        if(path.size() == nums.length){
            res.add(new ArrayList<>(path));
            return;
        }

        for(int i = 0; i < nums.length; ++i){
            if(used[i]) continue;
            if(i > 0 && nums[i] == nums[i -1] && !used[i - 1]) continue;

            used[i] = true;
            path.add(nums[i]);

            solve(res, nums, path, used);

            used[i] = false;
            path.removeLast();
        }
    }
}

File: 47. Permutations II.cpp
#include"LeetCodeLibs.h"



// Input: nums = [1,1,2]
// Output:
// [[1,1,2],
//  [1,2,1],
//  [2,1,1]]
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        vector<bool> used(nums.size());
        vector<int> path;
        sort(nums.begin(), nums.end());
    
        solve(res, nums, path, used);
        
        return res;
    }


    void solve(vector<vector<int>>& res, vector<int>& nums, vector<int>& path, vector<bool>& used){
        if(path.size() == nums.size()){
            res.push_back(path);
            return;
        }

        for(int i = 0; i < nums.size(); ++i){
            if (used[i]) continue;
            if(i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;
            used[i] = true;
            path.push_back(nums[i]);
            solve(res, nums, path, used);

            path.pop_back();
            used[i] = false;
        }
    }
    
};
